---
title: "IQS017 - Service actions raise errors when encountering failures"
---

IQS017 is as follows:

Service actions raise errors when encountering failures

## Reasoning

Things can go wrong when a service action is performed.
When this happens, the integration should raise an error to indicate that something went wrong.
These will be shown to the user in the UI, and can be used to help diagnose the issue.

## Example implementation

When the exception is caused by incorrect usage (for example incorrect input or referencing something that does not exist) should raise a `ServiceValidationError`.
When the exception is caused by an error in the service action itself (for example, a network error or a bug in the service), it should raise a `HomeAssistantError`.

In this example, we show a function that is registered as a service action in Home Assistant.
If the input is incorrect (when the end date is before the start date), a `ServiceValidationError` is raised, and if we can't reach the service, we raise a `HomeAssistantError`.

```python
async def async_set_schedule(call: ServiceCall) -> ServiceResponse:
    """Set the schedule for a day."""
    start_date = call.data[ATTR_START_DATE]
    end_date = call.data[ATTR_END_DATE]
    if end_date < start_date:
        raise ServiceValidationError("End date must be after start date")
    try:
        await client.set_schedule(start_date, end_date)
    except MyConnectionError as err:
        raise HomeAssistantError("Could not connect to the schedule") from err
```

## Additional resources

For more info on raising exceptions, check the [documentation](../../platform/raising_exceptions).

## Exceptions

There are no exceptions to this rule.

## Related rules

- IQS030 - Errors are translatable
- IQS035 - Services are added in async_setup
- IQS048 - The documentation described the provided actions that can be used.